File: 21_位运算/二进制枚举与三进制枚举/枚举子集/78. 子集.py
from itertools import chain, combinations
from typing import Any, Collection, Generator, List


# 1. dfs
class Solution1:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def dfs(index: int, path: List[int]):
            if index == len(nums):
                yield path[:]
                return
            path.append(nums[index])
            yield from dfs(index + 1, path)
            path.pop()
            yield from dfs(index + 1, path)

        return list(dfs(0, []))


# 2. powerset 顺序枚举
def powerset(collection: Collection[Any], isProperSubset=False):
    """求(真)子集,时间复杂度O(n*2^n)

    默认求所有子集
    """
    upper = len(collection) if isProperSubset else len(collection) + 1
    return chain.from_iterable(combinations(collection, n) for n in range(upper))


# 3. 枚举+check 顺序枚举
class Solution2:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def gen() -> Generator[List[int], None, None]:
            n = len(nums)
            for state in range(1 << n):
                cur = []
                for i in range(n):
                    if state & (1 << i):
                        cur.append(nums[i])
                yield cur

        return list(gen())


# 4. 滚动更新
class Solution3:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
        for num in nums:
            cur = []
            for pre in res:
                cur.append(pre + [num])
            res += cur
        return res

File: 21_位运算/二进制枚举与三进制枚举/枚举子集/78. 子集.ts
/**
 *
 * @param nums
 * @returns 2^n 个子集
 * @description 2^n*n时间复杂度
 */
function subsets<T>(nums: T[]): T[][] {
  const n = nums.length
  const res: T[][] = []

  for (let state = 0; state < 1 << n; state++) {
    const cands: T[] = []
    for (let j = 0; j < nums.length; j++) {
      if (state & (1 << j)) cands.push(nums[j])
    }

    res.push(cands)
  }

  return res
}

// 2^n 时间复杂度
function subsets2<T>(nums: T[]): T[][] {
  const n = nums.length
  const res: T[][] = []
  dfs(0, [])
  return res

  function dfs(index: number, path: T[]): void {
    if (index === n) {
      res.push(path.slice())
      return
    }

    dfs(index + 1, path)

    path.push(nums[index])
    dfs(index + 1, path)
    path.pop()
  }
}

function GosperHack(n: number, k: number) {
  let x = (1 << k) - 1
  const limit = 1 << n
  while (x < limit) {
    console.log(x.toString(2).padStart(5, '0'))
    const lowbit = x & -x
    const r = x + lowbit
    // xor
    x = r | (((x ^ r) >> 2) / lowbit)
  }
}

export {}

if (require.main === module) {
  console.log(subsets([1, 2, 3]))
  console.log(subsets2([1, 2, 3]))
}
